### PR TITLE
Add script to automate build of release artifacts

### PR DESCRIPTION
**Pull request by marun**
- Add script to build release artifacts

 - Remove documentation of artifact build (replaced by script)

 - Remove appVersion from chart yaml to avoid having to maintain it

Partial fix for #813.   I tried automating the tag creation/push and it ended up being more trouble than it was worth.

### CODE REVIEW

File: charts/federation-v2/Chart.yaml [hunk 1]
```diff
@@ -1,5 +1,4 @@
 apiVersion: v1
-appVersion: 0.0.2
```
Comment (xunpan) [inline charts/federation-v2/Chart.yaml]: any comments to remove this?

**Comment by xunpan** [comment]
Overall it LGTM.

File: charts/federation-v2/Chart.yaml (hunk 1 above)
Comment (marun) [reply charts/federation-v2/Chart.yaml]: Yes, it's in the first comment: `Remove appVersion from chart yaml to avoid having to maintain it`

Currently we don't have a reason to set an `appVersion` separate from the chart version, and `appVersion` is optional, so I figured it was preferable to remove it rather than having to update it with each release.

File: docs/releasing.md [hunk 2]
```diff
@@ -18,25 +18,14 @@ Creating a federation v2 release involves the following steps:
...
 4. Create github release
    1. Copy text from old release and replace old tag references
    2. Add a synopsis of the `Unreleased` section of `CHANGELOG.md`
    3. Add `kubefedctl.tgz` and `kubefedctl.tgz.sha`
    4. Add `federation-v2-<x.x.x>.tgz` and `federation-v2-<x.x.x>.tgz.sha`
 5. Update master
    1. Move the contents of the `Unreleased` section of `CHANGELOG.md` to `v<x.x.x>`
-   2. Propose a PR that includes changes to `charts/index.yaml` (from `4.6`) and `CHANGELOG.md`
+   2. Propose a PR that includes changes to `charts/index.yaml`
+      (updated by step 2) and `CHANGELOG.md`
```
Comment (font) [inline docs/releasing.md]: s/2/3/

File: scripts/build-release-artifacts.sh [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    helm package federation-v2
+
+    # Update the repo index (will need to be committed)
+    helm repo index . --merge index.yaml --url="https://github.com/kubernetes-sigs/federation-v2/releases/download/${RELEASE_TAG}"
+
+    sha256sum "federation-v2-${RELEASE_VERSION}.tgz" > "federation-v2-${RELEASE_VERSION}.tgz.sha"
+    mv federation-v2-${RELEASE_VERSION}.tgz* "${ROOT_DIR}/"
+
+    # Revert the chart changes (should not be committed)
+    mv federation-v2/values.yaml.backup federation-v2/values.yaml
```
Comment (font) [inline scripts/build-release-artifacts.sh]: Not part of this PR, but I'm thinking it might make sense to commit these changes as part of the tagged release, possibly once we go beta. The idea is that you can checkout a tagged release and deploy it without having to download any artifacts. This would probably be for dev use primarily.

Comment (marun) [reply scripts/build-release-artifacts.sh]: I don't know that what you're suggesting makes sense in the context of a 'tag'.  A branch, maybe, since that's a persistent context that can be updated, but a tag is just a persistent alias for a commit hash.

File: docs/releasing.md (hunk 2 above)
Comment (marun) [reply docs/releasing.md]: Done.

**Comment by marun** [comment]
@font Adding a tag and pushing that tag a) requires correct permissions b) can only be done once.  I opted to skip adding context-sensitive behavior and stick to just building the artifacts (in any case the only part of the release process that has proven tedious and potentially error-prone thus far).

Another manual step in the process that doesn't seem amenable to full automation - at least in the short-term - is the change log.  Maybe at some point we can move to scraping PR content as per the example of k/k, but for now the creation of a concise and user-focused changelog is a manual process taking both the changelog and the commit history as input.

File: scripts/build-release-artifacts.sh (hunk 3 above)
Comment (font) [reply scripts/build-release-artifacts.sh]: Right, and that commit object would point to code that contains the right version of stuff. It could either be based off `master` as we do today, but that would require a commit dance if we wanted to keep using the `canary` tag. Or a new/temporary branch named the same (or similarly) as the tag.

File: docs/releasing.md [hunk 4]
```diff
@@ -18,25 +18,14 @@ Creating a federation v2 release involves the following steps:
...
 4. Create github release
    1. Copy text from old release and replace old tag references
    2. Add a synopsis of the `Unreleased` section of `CHANGELOG.md`
    3. Add `kubefedctl.tgz` and `kubefedctl.tgz.sha`
    4. Add `federation-v2-<x.x.x>.tgz` and `federation-v2-<x.x.x>.tgz.sha`
 5. Update master
    1. Move the contents of the `Unreleased` section of `CHANGELOG.md` to `v<x.x.x>`
-   2. Propose a PR that includes changes to `charts/index.yaml` (from `4.6`) and `CHANGELOG.md`
+   2. Propose a PR that includes changes to `charts/index.yaml`
+      (updated by when the build script) and `CHANGELOG.md`
```
Comment (font) [inline docs/releasing.md]: s/when //

Comment (marun) [reply docs/releasing.md]: le sigh

**Comment by marun** [comment]
Updated

**Comment by shashidharatd** [comment]
@marun, are you planning to include building kubefedctl for macos in this pr or subsequent pr?

**Comment by marun** [comment]
@shashidharatd I'll add it to this PR.

File: scripts/build-release-artifacts.sh [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  >&2 echo "usage: $0 <release tag of the form v[0-9]+.[0-9]+.[0-9]+>"
+  exit 1
+fi
+
+ROOT_DIR="$(cd "$(dirname "$0")/.." ; pwd)"
+RELEASE_VERSION="${RELEASE_TAG:1:${#RELEASE_TAG}-1}"
+
+pushd "${ROOT_DIR}"
+  # Build release artifacts for kubefedctl
+  make kubefedctl
```
[the remark this answers is not in the document]
Comment (marun) [reply scripts/build-release-artifacts.sh]: As per other comments on this PR, the intention is to build binaries for both darwin and linux.

**Comment by marun** [comment]
The script has been updated to build kubefedctl artifacts for both darwin and linux, and the names of the tar files now include the version, distro and architecture.

File: scripts/build-release-artifacts.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  exit 1
+fi
+
+ROOT_DIR="$(cd "$(dirname "$0")/.." ; pwd)"
+RELEASE_VERSION="${RELEASE_TAG:1:${#RELEASE_TAG}-1}"
+
+pushd "${ROOT_DIR}"
+  # Build release artifacts for kubefedctl
+  make kubefedctl HOST_OS=linux
+  make kubefedctl HOST_OS=darwin
```
Comment (yamt) [inline scripts/build-release-artifacts.sh]: HOST_OS is intended to specify the target of the symlink.
as you don't need the symlink here, it's more straightforward to do `make bin/kubefedctl-linux bin/kubefedctl-darwin`

File: scripts/build-release-artifacts.sh [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  pushd "${ROOT_DIR}/bin"
+    # Building kubefedctl-[platform] automatically creates symlink to
+    # kubefectl.  Ensure the symlink points to the binary appropraite
```
Comment (yamt) [inline scripts/build-release-artifacts.sh]: s/kubefectl/kubefedctl/

File: scripts/build-release-artifacts.sh [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    # to the platform.
+    #
+    # TODO(marun) Fix the makefile to only create the symlink if the
+    # binary suffix matches the host os.
+    HOST_OS="$(uname -s | tr A-Z a-z)"
+    ln -sf "kubefedctl-${HOST_OS}" kubefedctl
```
Comment (yamt) [inline scripts/build-release-artifacts.sh]: i couldn't find what needs this symlink. is it necessary?

**Comment by marun** [comment]
@yamt Thanks for the pointers, I've updated accordingly.

**Comment by shashidharatd** [comment]
/lgtm
Thanks !
